File: frontend/src/types/game.ts
export interface Genre {
  id: number;
  name: string;
  slug: string;
}

export interface Platform {
  id: number;
  name: string;
  slug: string;
}

export interface Tag {
  id: number;
  name: string;
  slug: string;
}

export interface Screenshot {
  id: number;
  image: string;
}

/** Lightweight — used in lists and catalog */
export interface GameListItem {
  id: number;
  name: string;
  slug: string;
  released: string | null;
  background_image: string | null;
  rating: number | null;
  genres: Genre[];
  platforms: Platform[];
  playtime: number | null;
  hltb_main_hours: number | null;
}

/** Full detail — used on the game detail page */
export interface Game extends GameListItem {
  rawg_id: number;
  description: string | null;
  ratings_count: number;
  metacritic: number | null;
  tags: Tag[];
  screenshots: Screenshot[];
  hltb_main_hours: number | null;
  hltb_main_extra_hours: number | null;
  hltb_completionist_hours: number | null;
}

export interface PaginatedGames {
  total: number;
  page: number;
  page_size: number;
  results: GameListItem[];
}

export interface GameFilters {
  search?: string;
  genre?: string;
  platform?: string;
  year?: number;
  min_rating?: number;
  max_hours?: number;
  library_state?: 'all' | 'saved' | 'not_saved';
  sort?: 'rating_desc' | 'released_desc' | 'name_asc' | 'playtime_asc';
}

File: backend/app/schemas/game.py
from datetime import date
from typing import Any

from pydantic import BaseModel


class GameOut(BaseModel):
    """Full game detail — used on the game detail page."""
    id:               int
    rawg_id:          int
    name:             str
    slug:             str
    description:      str | None
    released:         date | None
    background_image: str | None
    rating:           float | None
    ratings_count:    int
    metacritic:       int | None
    genres:           list[Any]
    platforms:        list[Any]
    tags:             list[Any]
    screenshots:      list[Any]

    model_config = {"from_attributes": True}


class GameListOut(BaseModel):
    """Lightweight game info — used in list/catalog views and nested in other schemas."""
    id:               int
    name:             str
    slug:             str
    released:         date | None
    background_image: str | None
    rating:           float | None
    genres:           list[Any]
    platforms:        list[Any]

    model_config = {"from_attributes": True}


class PaginatedGames(BaseModel):
    total:     int
    page:      int
    page_size: int
    results:   list[GameListOut]
